core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -392,9 +392,23 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
     // Remove all the lingering executors that should be removed but not yet. The reason might be
     // because (1) disconnected event is not yet received; (2) executors die silently.
-    executorDataMap.toMap.foreach { case (eid, _) =>
-      driverEndpoint.askWithRetry[Boolean](
-        RemoveExecutor(eid, SlaveLost("Stale executor after cluster manager re-registered.")))
+    // Note: here copy the code of remove executor from DriverEndpoint to avoid deadlock(reset
+    // and removeExecutor both to get the lock of schedulerbackend.)
+    val reason = SlaveLost("Stale executor after cluster manager re-registered.")
+    executorDataMap.toMap.foreach { case (executorId, executorInfo) =>
```

Comment:
`executorDataMap` should not be modified outside DriverEndpoint. See the comment for `executorDataMap`.